zebra/zebra_evpn_neigh.c

```diff
@@ -0,0 +1,2452 @@
+}
+
+struct hash *zebra_neigh_db_create(const char *desc)
+{
+	return hash_create(neigh_hash_keymake, neigh_cmp, desc);
+}
+
+uint32_t num_dup_detected_neighs(zebra_evpn_t *zevpn)
+{
+	unsigned int i;
+	uint32_t num_neighs = 0;
+	struct hash *hash;
+	struct hash_bucket *hb;
+	zebra_neigh_t *nbr;
+
+	hash = zevpn->neigh_table;
+	if (!hash)
+		return num_neighs;
+	for (i = 0; i < hash->size; i++) {
+		for (hb = hash->index[i]; hb; hb = hb->next) {
+			nbr = (zebra_neigh_t *)hb->data;
+			if (CHECK_FLAG(nbr->flags, ZEBRA_NEIGH_DUPLICATE))
+				num_neighs++;
+		}
+	}
+
+	return num_neighs;
+}
+
+/*
+ * Helper function to determine maximum width of neighbor IP address for
+ * display - just because we're dealing with IPv6 addresses that can
+ * widely vary.
+ */
+void zebra_evpn_find_neigh_addr_width(struct hash_bucket *bucket, void *ctxt)
+{
+	zebra_neigh_t *n;
+	char buf[INET6_ADDRSTRLEN];
+	struct neigh_walk_ctx *wctx = ctxt;
+	int width;
+
+	n = (zebra_neigh_t *)bucket->data;
+
+	ipaddr2str(&n->ip, buf, sizeof(buf));
+	width = strlen(buf);
+	if (width > wctx->addr_width)
+		wctx->addr_width = width;
+}
+
+/*
+ * Count of remote neighbors referencing this MAC.
+ */
+int remote_neigh_count(zebra_mac_t *zmac)
+{
+	zebra_neigh_t *n = NULL;
+	struct listnode *node = NULL;
+	int count = 0;
+
+	for (ALL_LIST_ELEMENTS_RO(zmac->neigh_list, node, n)) {
+		if (CHECK_FLAG(n->flags, ZEBRA_NEIGH_REMOTE))
+			count++;
+	}
+
+	return count;
+}
+
+/*
+ * Install remote neighbor into the kernel.
+ */
+int zebra_evpn_rem_neigh_install(zebra_evpn_t *zevpn, zebra_neigh_t *n,
+				 bool was_static)
+{
+	struct interface *vlan_if;
+	int flags;
+	int ret = 0;
+
+	if (!(n->flags & ZEBRA_NEIGH_REMOTE))
+		return 0;
+
+	vlan_if = zevpn_map_to_svi(zevpn);
+	if (!vlan_if)
+		return -1;
+
+	flags = DPLANE_NTF_EXT_LEARNED;
+	if (n->flags & ZEBRA_NEIGH_ROUTER_FLAG)
+		flags |= DPLANE_NTF_ROUTER;
+	ZEBRA_NEIGH_SET_ACTIVE(n);
+
+	dplane_rem_neigh_add(vlan_if, &n->ip, &n->emac, flags, was_static);
+
+	return ret;
+}
+
+/*
+ * Install neighbor hash entry - called upon access VLAN change.
+ */
+void zebra_evpn_install_neigh_hash(struct hash_bucket *bucket, void *ctxt)
+{
+	zebra_neigh_t *n;
+	struct neigh_walk_ctx *wctx = ctxt;
+
+	n = (zebra_neigh_t *)bucket->data;
+
+	if (CHECK_FLAG(n->flags, ZEBRA_NEIGH_REMOTE))
+		zebra_evpn_rem_neigh_install(wctx->zevpn, n,
+					     false /*was_static*/);
+}
+
+/*
+ * Callback to allocate neighbor hash entry.
+ */
+static void *zebra_evpn_neigh_alloc(void *p)
+{
+	const zebra_neigh_t *tmp_n = p;
+	zebra_neigh_t *n;
+
+	n = XCALLOC(MTYPE_NEIGH, sizeof(zebra_neigh_t));
+	*n = *tmp_n;
+
+	return ((void *)n);
+}
+
+static void zebra_evpn_local_neigh_ref_mac(zebra_neigh_t *n,
+					   struct ethaddr *macaddr,
+					   zebra_mac_t *mac,
+					   bool send_mac_update)
+{
+	char macbuf[ETHER_ADDR_STRLEN];
+	char ipbuf[INET6_ADDRSTRLEN];
+	bool old_static;
+	bool new_static;
+
+	memcpy(&n->emac, macaddr, ETH_ALEN);
```

Comment:
Static analyser is complaining about possible `macaddr` `NULL` here. If you think it will never be `NULL` you could add an `assert` to fix it.

Also I think you should move this block to after `if (!mac) return` otherwise it is wasted work and prone to static analyser warnings.